phpBB/phpbb/style/manager.php

```diff
@@ -0,0 +1,436 @@
+		foreach ($installed as $style)
+		{
+			$installed_dirs[] = $style['style_path'];
+			$installed_names[$style['style_name']] = array(
+				'path'		=> $style['style_path'],
+				'id'		=> $style['style_id'],
+				'parent'	=> $style['style_parent_id'],
+				'tree'		=> (strlen($style['style_parent_tree']) ? $style['style_parent_tree'] . '/' : '') . $style['style_path'],
+			);
+		}
+
+		// Get list of directories
+		$dirs = $this->find_style_dirs();
+
+		// Find styles that can be installed
+		$styles = array();
+		foreach ($dirs as $dir)
+		{
+			if (in_array($dir, $installed_dirs))
+			{
+				// Style is already installed
+				continue;
+			}
+			$cfg = $this->read_style_cfg($dir);
+			if ($cfg === false)
+			{
+				// Invalid style.cfg
+				continue;
+			}
+
+			// Style should be available for installation
+			$parent = $cfg['parent'];
+			$style = array(
+				'style_id'			=> 0,
+				'style_name'		=> $cfg['name'],
+				'style_copyright'	=> $cfg['copyright'],
+				'style_active'		=> 0,
+				'style_path'		=> $dir,
+				'bbcode_bitfield'	=> $cfg['template_bitfield'],
+				'style_parent_id'	=> 0,
+				'style_parent_tree'	=> '',
+				// Extra values for styles list
+				// All extra variable start with _ so they won't be confused with data that can be added to styles table
+				'_inherit_name'			=> $parent,
+				'_available'			=> true,
+				'_note'					=> '',
+			);
+
+			// Check style inheritance
+			if ($parent != '')
+			{
+				if (isset($installed_names[$parent]))
+				{
+					// Parent style is installed
+					$row = $installed_names[$parent];
+					$style['style_parent_id'] = $row['id'];
+					$style['style_parent_tree'] = $row['tree'];
+				}
+				else
+				{
+					// Parent style is not installed yet
+					$style['_available'] = false;
+					$style['_note'] = sprintf($this->user->lang['REQUIRES_STYLE'], htmlspecialchars($parent));
+				}
+			}
+
+			if ($all || $style['_available'])
+			{
+				$styles[] = $style;
+			}
+		}
+
+		return $styles;
+	}
+
+	/**
+	* Find all directories that have styles
+	*
+	* @return array Directory names
+	*/
+	protected function find_style_dirs()
+	{
+		$styles = array();
+
+		$dp = @opendir($this->styles_path);
+		if ($dp)
+		{
+			while (($file = readdir($dp)) !== false)
+			{
+				$dir = $this->styles_path . $file;
+				if ($file[0] == '.' || !is_dir($dir))
+				{
+					continue;
+				}
+
+				if (file_exists("{$dir}/style.cfg"))
+				{
+					$styles[] = $file;
+				}
+			}
+			closedir($dp);
+		}
+
+		return $styles;
+	}
+
+	/**
+	* Lists all styles
+	*
+	* @return array Rows with styles data
+	*/
+	public function get_installed_styles()
+	{
+		$sql = 'SELECT *
+			FROM ' . $this->styles_table;
+		$result = $this->db->sql_query($sql);
+
+		$rows = $this->db->sql_fetchrowset($result);
+		$this->db->sql_freeresult($result);
+
+		return $rows;
+	}
+
+	public function get_style_data($field, $value)
+	{
+		// TODO: Review this, possible security issue
+		// if not, maybe field doesnt need to be escaped
```

Comment:
Any escaped sstring will be invalid and may be exploited. I think there is 3 solutions:
 - not escape 
 - compare the escaped string and not escaped one and throw if they are not identical
 - checks against a regular expression 

What was done in the original method?

@marc1706